### PR TITLE
Add WebAssembly package with WASM binary writer and reader

### PR DESCRIPTION
**Pull request by turbolent**
Start work on experimenting with WebAssembly: Add a new package which allows writing and reading WASM binaries. For now, only enough instructions are implemented to implement an add function, i.e. `local.get`, `i32.add`, and `end`

### CODE REVIEW

File: runtime/compiler/wasm/reader_test.go [hunk 1]
```diff
@@ -0,0 +1,825 @@
...
+func TestWasmReader_readMagicAndVersion(t *testing.T) {
+
+	t.Parallel()
+
+	read := func(data []byte) error {
+		b := buf{data: data}
+		r := WASMReader{buf: &b}
+		return r.readMagicAndVersion()
+	}
+
```
Comment (zhangchiqing) [inline runtime/compiler/wasm/reader_test.go:38]: Missing happy cases:  "valid magic" and "valid magic and version"

File: runtime/compiler/wasm/reader_test.go [hunk 2]
```diff
@@ -0,0 +1,825 @@
...
+					Offset:  8,
+					ValType: 0xFF,
+				},
+			},
+			err,
+		)
+		assert.Nil(t, funcTypes)
+	})
+
+	t.Run("invalid, parameter type missing", func(t *testing.T) {
```
Comment (zhangchiqing) [inline runtime/compiler/wasm/reader_test.go:332]: 👍 

This is a case when not enough parameter type is provided, I wonder if too many parameter types are provided, what error would it return? I guess the extra parameter type would be parsed as the next item, which is "return value count", which would shows a different error message?

File: runtime/compiler/wasm/reader_test.go [hunk 3]
```diff
@@ -0,0 +1,825 @@
...
+		assert.Nil(t, funcTypes)
+	})
+
+	t.Run("invalid count", func(t *testing.T) {
+
+		t.Parallel()
+
+		funcTypes, err := read([]byte{
+			// section size: 7 (LEB128)
+			0x80, 0x80, 0x80, 0x80, 0x0,
```
Comment (zhangchiqing) [inline runtime/compiler/wasm/reader_test.go:511]: first byte for `section size: 7` should be `0x87`, but got `0x80` here

File: runtime/compiler/wasm/leb128_test.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+	t.Run("write: max byte count", func(t *testing.T) {
+		var b buf
+		err := b.writeULEB128(math.MaxUint32)
+		require.NoError(t, err)
+		require.GreaterOrEqual(t, max32bitLEB128ByteCount, len(b.data))
+	})
+
+	t.Run("read: max byte count", func(t *testing.T) {
+		b := buf{data: []byte{0x81, 0x82, 0x83, 0x84, 0x85, 0x86, 0x87, 0x88}}
```
Comment (zhangchiqing) [inline runtime/compiler/wasm/leb128_test.go:63]: Do these bytes represent the biggest value can be read?
isn't biggest uint32 can be represented as `0xFF, 0xFF, 0xFF, 0xFF`?

File: runtime/compiler/wasm/writer_test.go [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+	require.Equal(t,
+		`(module
+  (type (;0;) (func (param i32 i32) (result i32)))
+  (func (;0;) (type 0) (param i32 i32) (result i32)
+    (local i32)
+    local.get 0
+    local.get 1
+    i32.add))
+`,
+		wasm2wat(b.data),
```
Comment (zhangchiqing) [inline runtime/compiler/wasm/writer_test.go:286]: 👍  Nice to compare to the official library

File: runtime/compiler/wasm/reader_test.go [hunk 6]
```diff
@@ -0,0 +1,825 @@
...
+
+	t.Run("valid", func(t *testing.T) {
+
+		t.Parallel()
+
+		codes, err := read([]byte{
+			// section size: 15 (LEB128)
+			0x8f, 0x80, 0x80, 0x80, 0x0,
+			// function count: 1
+			0x1,
```
Comment (zhangchiqing) [inline runtime/compiler/wasm/reader_test.go:569]: I was afraid these raw bytes would be unreadable, and useless for review. 
But after reading the code, these bytes and comments are actually quite easier to read!
Nice comments 👍

**Review by zhangchiqing — APPROVED** [review]
As always, very clean implementation and great tests!

I had comments and questions, but looks great 👍

**Comment by turbolent** [comment]
@zhangchiqing Thank you for the review! 

I added documentation as a package comment in 40ccbda. Hopefully that answers some of your questions, I'll add more in the other packages later. Let me know if you still have questions regarding the `wasm` package and I can clarify.

File: runtime/compiler/wasm/buf.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+		buf.data = append(buf.data, b)
+	}
+	buf.offset++
+	return nil
+}
+
+func (buf *buf) WriteBytes(data []byte) error {
+	for _, b := range data {
+		err := buf.WriteByte(b)
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/compiler/wasm/buf.go:47]: `WriteByte` implements the `io.ByteWriter` interface, which requires `WriteByte(c byte) error`

File: runtime/compiler/wasm/magic.go [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package wasm
+
+// wasmMagic is the magic byte sequence that appears at the start of the WASM binary
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/compiler/wasm/magic.go]: Good idea, done in 43a4951

File: runtime/compiler/wasm/reader_test.go (hunk 3 above)
Comment (turbolent) [reply runtime/compiler/wasm/reader_test.go:511]: Good catch, fixed in ebf8166

File: runtime/compiler/wasm/section.go [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+package wasm
+
+// sectionID is the ID of a section in the WASM binary
+//
+type sectionID byte
+
+const (
+	sectionIDType     sectionID = 1
+	sectionIDFunction sectionID = 3
+	sectionIDCode     sectionID = 10
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/compiler/wasm/section.go:45]: Good idea, done in 6246b9e

File: runtime/compiler/wasm/writer_test.go (hunk 5 above)
Comment (turbolent) [reply runtime/compiler/wasm/writer_test.go:286]: Note that the tool is not called to "compare": The wasm package produces a binary, and the `wasm2wat` tool is used to check if the written binary (WASM) is valid by parsing it and converting it to the text format (WAT).

File: runtime/compiler/wasm/reader_test.go (hunk 6 above)
Comment (turbolent) [reply runtime/compiler/wasm/reader_test.go:569]: I'm glad they're helpful! I think once you have internalized the structure of the binary format, it becomes readable and when something is mis-produced, it helps with debugging.

File: runtime/compiler/wasm/leb128_test.go (hunk 4 above)
Comment (turbolent) [reply runtime/compiler/wasm/leb128_test.go:63]: No they don't, note it says "max byte count" not "max byte value". This test ensures that only up to the maximum number of bytes are read when reading a LEB128-encoded 32-bit number, which is 5 (see https://github.com/onflow/cadence/blob/ea6b35ba72b9622a814c3d27d718d120a594f06f/runtime/compiler/wasm/leb128.go#L145-L151), i.e. the function checks that only 5 out of the 8 bytes are read, to ensure the LEB128 parser doesn't keep reading infinitely.

I've added comments to clarify this in 4e7e058c9fd16e4884825d3db3cf9b768b70eb3a

File: runtime/compiler/wasm/reader_test.go (hunk 1 above)
Comment (turbolent) [reply runtime/compiler/wasm/reader_test.go:38]: See below, https://github.com/onflow/cadence/pull/366/files#diff-0f55d4941b179be22cb3099b9e25051d6d10cf2bf71fd597d8d50d0007bea6ebR99

**Pull request merged**